cgen error, when compiling `mut r3 := Zip[int,int]{}`, for `struct Zip[A,B] { a Iterator[A] b Iterator[B] }`

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.9 6b0c272, press to see full `v doctor` output</summary>

|V full version      |V 0.4.9 bd10e12.6b0c272
|:-------------------|:-------------------
|OS                  |linux, Ubuntu 20.04.6 LTS
|Processor           |4 cpus, 64bit, little endian, Intel(R) Core(TM) i3-3225 CPU @ 3.30GHz
|Memory              |3.3GB/15.05GB
|                    |
|V executable        |/home/delian/code/v/v
|V last modified time|2025-01-23 09:58:28
|                    |
|V home dir          |OK, value: /home/delian/code/v
|VMODULES            |OK, value: /home/delian/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/delian/code/v
|                    |
|Git version         |git version 2.47.1
|V git status        |weekly.2025.03-25-g6b0c2722
|.git/config present |true
|                    |
|cc version          |cc (Ubuntu 10.5.0-1ubuntu1~20.04) 10.5.0
|gcc version         |gcc (Ubuntu 10.5.0-1ubuntu1~20.04) 10.5.0
|clang version       |clang version 10.0.0-4ubuntu1 
|tcc version         |tcc version 0.9.28rc 2024-07-31 HEAD@1cee0908 (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 0134e9b9
|emcc version        |N/A
|glibc version       |ldd (Ubuntu GLIBC 2.31-0ubuntu9.16) 2.31

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  /home/delian/r.v && /home/delian/r`
```v
pub struct Range[T] {
pub:
	from T
	to T
	inclusive bool
pub mut:	
	step T = T(1)
	i T
mut:	
	started bool
}

pub fn (mut r Range[T]) next[T]() ?T {
	if !r.started {
		r.started = true
		assert r.step > 0
		if r.from > r.to {
			r.step = -r.step
		}
		r.i = r.from
	}
	i := r.i
	next_i := i + r.step
	if r.inclusive {
		if r.step < 0 && i < r.to {
			return none
		}
		if r.step > 0 && i > r.to {
			return none
		}
	} else {
		if r.step < 0 && i <= r.to {
			return none
		}
		if r.step > 0 && i >= r.to {
			return none
		}
	}
	r.i = next_i
	return i
}


pub interface Iterator[T] {
mut:
	next() ?T
}

pub struct Zip[A,B] {
mut:
	a Iterator[A]
	b Iterator[B]
}
pub struct Pair[A,B] {
	a A
	b B
}

pub fn (mut z Zip[A,B]) next[A,B]() ?Pair[A,B] {
	a := z.a.next()?
	b := z.b.next()?
	return Pair[A,B]{a,b}
}

for i in Range{from: 5, to: 10} { println('from 5 to 10, i= ${i}') }
for i in Range{from: 5, to: 10, inclusive: true} { println('from 5 to 10, inclusive, i= ${i}') }
for i in Range{from: 10, to: 5, inclusive: true} { println('from 10 to 5, inclusive, i= ${i}') }
for i in Range{from: 10, to: 5} { println('from 10 to 5, i= ${i}') }

for x in Range{from: 2.0, to: 3.0, step: 0.5} { println('from 2.0 to 3.0, step: 0.5, i= ${x}') }
for x in Range{from: 2.0, to: 3.0, step: 0.5, inclusive: true} { println('from 2.0 to 3.0, step: 0.5, inclusive, i= ${x}') }

mut r1 := Range{from: 5, to: 10}
mut i1 := Iterator[int](r1)
println(i1)
mut r2 := Range{from: 10, to: 5}
mut i2 := Iterator[int](r2)
println(i2)

mut r3 := Zip[int,int]{} // cgen error for main__Iterator_T_int
//r3.a = i1 // checker error: cannot implement interface `Iterator<A>[int]` with a different interface `Iterator[int]`
//r3.b = i2 // checker error: cannot implement interface `Iterator<A>[int]` with a different interface `Iterator[int]`
//dump(r3)
//for x in r3 {
//	println(x)
//}

```

**What did you see?**
```
/home/delian/r.v:80:11: notice: interface field `Zip[int, int].a` must be initialized
   78 | println(i2)
   79 | 
   80 | mut r3 := Zip[int,int]{} // cgen error for main__Iterator_T_int
      |           ~~~~~~~~~~~~~~
   81 | //r3.a = i1 // checker error: cannot implement interface `Iterator<A>[int]` with a different interface `Iterator[int]`
   82 | //r3.b = i2 // checker error: cannot implement interface `Iterator<A>[int]` with a different interface `Iterator[int]`
/home/delian/r.v:80:11: notice: interface field `Zip[int, int].b` must be initialized
   78 | println(i2)
   79 | 
   80 | mut r3 := Zip[int,int]{} // cgen error for main__Iterator_T_int
      |           ~~~~~~~~~~~~~~
   81 | //r3.a = i1 // checker error: cannot implement interface `Iterator<A>[int]` with a different interface `Iterator[int]`
   82 | //r3.b = i2 // checker error: cannot implement interface `Iterator<A>[int]` with a different interface `Iterator[int]`
/home/delian/r.v:80:5: warning: unused variable: `r3`
   78 | println(i2)
   79 | 
   80 | mut r3 := Zip[int,int]{} // cgen error for main__Iterator_T_int
      |     ~~
   81 | //r3.a = i1 // checker error: cannot implement interface `Iterator<A>[int]` with a different interface `Iterator[int]`
   82 | //r3.b = i2 // checker error: cannot implement interface `Iterator<A>[int]` with a different interface `Iterator[int]`
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/r.01JJ99K99KTPH0SK41R6JFASP5.tmp.c:861: error: struct/union/enum already defined
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
```

**What did you expect to see?**

a compiled program

--- a/vlib/v/ast/table.v
+++ b/vlib/v/ast/table.v
@@ -2190,16 +2190,15 @@ pub fn (mut t Table) unwrap_generic_type_ex(typ Type, generic_names []string, co
 		Interface {
 			// resolve generic types inside methods
 			mut imethods := ts.info.methods.clone()
+			gn_names := t.get_real_generic_names(typ, generic_names)
 			for mut method in imethods {
-				if unwrap_typ := t.convert_generic_type(method.return_type, generic_names,
-					concrete_types)
+				if unwrap_typ := t.convert_generic_type(method.return_type, gn_names,
+					concrete_types[..gn_names.len])
 				{
 					method.return_type = unwrap_typ
 				}
 				for mut param in method.params {
-					if unwrap_typ := t.convert_generic_type(param.typ, generic_names,
-						concrete_types)
-					{
+					if unwrap_typ := t.convert_generic_type(param.typ, gn_names, concrete_types) {
 						param.typ = unwrap_typ
 					}
 				}
@@ -2289,6 +2288,7 @@ pub fn (mut t Table) generic_insts_to_concrete() {
 									fields[i].typ = t.unwrap_generic_type(fields[i].typ,
 										generic_names, info.concrete_types)
 								}
+
 								if t_typ := t.convert_generic_type(fields[i].typ, generic_names,
 									info.concrete_types)
 								{
@@ -2462,6 +2462,18 @@ pub fn (mut t Table) generic_insts_to_concrete() {
 			}
 		}
 	}
+}
+
+// Extracts all generic names from Type<A>[B] => B when <A>[B] is present
+// otherwise generic_names is returned
+pub fn (t &Table) get_real_generic_names(typ Type, generic_names []string) []string {
+	if typ.has_flag(.generic) {
+		typ_name := t.type_to_str(typ)
+		if typ_name.contains('>[') {
+			return typ_name.split('>[')[1].all_before_last(']').split(',')
+		}
+	}
+	return generic_names
 }
 
 // Extracts all type names from given types, notice that MultiReturn will be decompose

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -1107,6 +1107,7 @@ pub fn (mut g Gen) get_sumtype_variant_name(typ ast.Type, sym ast.TypeSymbol) st
 pub fn (mut g Gen) write_typeof_functions() {
 	g.writeln('')
 	g.writeln('// >> typeof() support for sum types / interfaces')
+	mut already_generated_ifaces := map[string]bool{}
 	for ityp, sym in g.table.type_symbols {
 		if sym.kind == .sum_type {
 			if g.pref.skip_unused && sym.idx !in g.table.used_features.used_syms {
@@ -1177,6 +1178,10 @@ pub fn (mut g Gen) write_typeof_functions() {
 			if g.pref.skip_unused && sym.idx !in g.table.used_features.used_syms {
 				continue
 			}
+			if sym.cname in already_generated_ifaces {
+				continue
+			}
+			already_generated_ifaces[sym.cname] = true
 			g.definitions.writeln('${g.static_non_parallel}char * v_typeof_interface_${sym.cname}(int sidx);')
 			if g.pref.parallel_cc {
 				g.extern_out.writeln('extern char * v_typeof_interface_${sym.cname}(int sidx);')
@@ -1805,11 +1810,15 @@ static inline void __${sym.cname}_pushval(${sym.cname} ch, ${push_arg} val) {
 			interface_non_generic_syms << sym
 		}
 	}
+	mut already_generated_ifaces := map[string]bool{}
 	for sym in interface_non_generic_syms {
 		if g.pref.skip_unused && sym.idx !in g.table.used_features.used_syms {
 			continue
 		}
-		g.write_interface_typesymbol_declaration(sym)
+		if sym.cname !in already_generated_ifaces {
+			g.write_interface_typesymbol_declaration(sym)
+			already_generated_ifaces[sym.cname] = true
+		}
 	}
 }
 
@@ -7859,6 +7868,7 @@ fn (mut g Gen) interface_table() string {
 	}
 	mut sb := strings.new_builder(100)
 	mut conversion_functions := strings.new_builder(100)
+	mut already_generated_ifaces := map[string]bool{}
 	interfaces := g.table.type_symbols.filter(it.kind == .interface && it.info is ast.Interface)
 	for isym in interfaces {
 		inter_info := isym.info as ast.Interface
@@ -7868,6 +7878,10 @@ fn (mut g Gen) interface_table() string {
 		if g.pref.skip_unused && isym.idx !in g.table.used_features.used_syms {
 			continue
 		}
+		if isym.cname in already_generated_ifaces {
+			continue
+		}
+		already_generated_ifaces[isym.cname] = true
 		// interface_name is for example Speaker
 		interface_name := isym.cname
 		// generate a struct that references interface methods

--- a/vlib/v/markused/walker.v
+++ b/vlib/v/markused/walker.v
@@ -1295,9 +1295,6 @@ fn (mut w Walker) mark_resource_dependencies() {
 		w.fn_by_name('strings.new_builder')
 		w.uses_free[ast.string_type] = true
 	}
-	if w.uses_eq {
-		w.fn_by_name('fast_string_eq')
-	}
 	if w.features.auto_str_ptr {
 		w.fn_by_name('isnil')
 		w.fn_by_name('tos4')
@@ -1573,7 +1570,6 @@ pub fn (mut w Walker) finalize(include_panic_deps bool) {
 	}
 	if (w.used_option + w.used_result + w.used_none) > 0 {
 		w.mark_const_as_used('none__')
-		w.fn_by_name('memdup')
 	}
 	if include_panic_deps || w.uses_external_type || w.uses_asserts || w.uses_debugger
 		|| w.uses_interp {
@@ -1592,9 +1588,11 @@ pub fn (mut w Walker) finalize(include_panic_deps bool) {
 		w.fn_by_name(ref_array_idx_str + '.push')
 		w.fn_by_name(string_idx_str + '.substr')
 		w.fn_by_name('v_fixed_index')
-		w.fn_by_name('fast_string_eq')
 		w.mark_by_sym_name('StrIntpData')
 		w.mark_by_sym_name('StrIntpMem')
+	}
+	if w.uses_eq {
+		w.fn_by_name('fast_string_eq')
 	}
 
 	// remove unused symbols

--- a/vlib/v/markused/walker.v
+++ b/vlib/v/markused/walker.v
@@ -1573,6 +1573,7 @@ pub fn (mut w Walker) finalize(include_panic_deps bool) {
 	}
 	if (w.used_option + w.used_result + w.used_none) > 0 {
 		w.mark_const_as_used('none__')
+		w.fn_by_name('memdup')
 	}
 	if include_panic_deps || w.uses_external_type || w.uses_asserts || w.uses_debugger
 		|| w.uses_interp {
@@ -1591,6 +1592,7 @@ pub fn (mut w Walker) finalize(include_panic_deps bool) {
 		w.fn_by_name(ref_array_idx_str + '.push')
 		w.fn_by_name(string_idx_str + '.substr')
 		w.fn_by_name('v_fixed_index')
+		w.fn_by_name('fast_string_eq')
 		w.mark_by_sym_name('StrIntpData')
 		w.mark_by_sym_name('StrIntpMem')
 	}

--- a/vlib/v/tests/generics/generic_interface_field_test.v
+++ b/vlib/v/tests/generics/generic_interface_field_test.v
@@ -1,0 +1,79 @@
+pub struct Range[T] {
+pub:
+	from      T
+	to        T
+	inclusive bool
+pub mut:
+	step T = T(1)
+	i    T
+mut:
+	started bool
+}
+
+pub fn (mut r Range[T]) next[T]() ?T {
+	if !r.started {
+		r.started = true
+		assert r.step > 0
+		if r.from > r.to {
+			r.step = -r.step
+		}
+		r.i = r.from
+	}
+	i := r.i
+	next_i := i + r.step
+	if r.inclusive {
+		if r.step < 0 && i < r.to {
+			return none
+		}
+		if r.step > 0 && i > r.to {
+			return none
+		}
+	} else {
+		if r.step < 0 && i <= r.to {
+			return none
+		}
+		if r.step > 0 && i >= r.to {
+			return none
+		}
+	}
+	r.i = next_i
+	return i
+}
+
+pub interface Iterator[T] {
+mut:
+	next() ?T
+}
+
+pub struct Zip[A, B] {
+mut:
+	a Iterator[A]
+	b Iterator[B]
+}
+
+pub struct Pair[A, B] {
+	a A
+	b B
+}
+
+pub fn (mut z Zip[A, B]) next[A, B]() ?Pair[A, B] {
+	a := z.a.next()?
+	b := z.b.next()?
+	return Pair[A, B]{a, b}
+}
+
+fn test_main() {
+	mut r1 := Range{
+		from: 5
+		to:   10
+	}
+	_ := Iterator[int](r1)
+
+	mut r2 := Range{
+		from: 10
+		to:   5
+	}
+	_ := Iterator[int](r2)
+
+	_ := Zip[int, int]{}
+}

